controllers/hvpa_controller.go

```diff
@@ -1266,3 +989,273 @@ func getThreshold(thresholdVals *autoscalingv1alpha1.ChangeThreshold, resourceTy
+			// Get latest HVPA object
+			hvpa := &autoscalingv1alpha1.Hvpa{}
+			err := client.Get(context.TODO(), types.NamespacedName{Name: name, Namespace: a.Meta.GetNamespace()}, hvpa)
+			if err != nil {
+				log.Error(err, "Error retreiving hvpa")
+				return nil
+			}
+			clone := hvpa.DeepCopy()
+
+			hvpaStatus := clone.Status
+			if hvpaStatus.LastScaling.LastScaleTime == nil || hvpaStatus.OverrideScaleUpStabilization == true {
+				log.V(4).Info("HVPA status already set to override last scale time.")
+				return nil
+			}
+
+			if pod.Status.Reason == "Evicted" {
+				// If pod was evicted beause of 'KubeletHasInsufficientMemory' node condition,
+				// only then we want to continue, otherwise exit
+				req := types.NamespacedName{
+					Name: nodeName,
+				}
+				node := corev1.Node{}
+				err := client.Get(context.TODO(), req, &node)
+				if err != nil {
+					log.Error(err, "Error fetching node")
+					return nil
+				}
+
+				hasMemPressure := false
+				for _, condition := range node.Status.Conditions {
+					if condition.Reason == "KubeletHasInsufficientMemory" && condition.Status == corev1.ConditionTrue {
+						hasMemPressure = true
+						break
+					}
+				}
+				if !hasMemPressure {
+					log.V(4).Info("Pod was evicted, but the node is not under memory pressure.")
+					return nil
+				}
+				log.V(4).Info("Pod might have been evited because node was under memory pressure")
+			} else {
+				// The pod was oomkilled
+				// Check if scaling already happened after this was oomkilled
+				recent := false
+				for i := range pod.Status.ContainerStatuses {
+					containerStatus := &pod.Status.ContainerStatuses[i]
+					if containerStatus.RestartCount > 0 &&
+						containerStatus.LastTerminationState.Terminated != nil &&
+						containerStatus.LastTerminationState.Terminated.Reason == "OOMKilled" &&
+						clone.Status.LastScaling.LastScaleTime != nil &&
+						containerStatus.LastTerminationState.Terminated.FinishedAt.After(clone.Status.LastScaling.LastScaleTime.Time) {
+
+						recent = true
+						break
+					}
+				}
+				if !recent {
+					log.V(4).Info("This is not  a recent oomkill. Return")
+					return nil
+				}
+			}
+
+			clone.Status.OverrideScaleUpStabilization = true
+
+			log.V(2).Info("Updating HVPA status to override last scale time", "HVPA", clone.Name)
+			err = client.Update(context.TODO(), clone)
+			if err != nil {
+				log.Error(err, "Error overrinding last scale time for", "HVPA", name)
+			}
+
+			return nil
+		}),
+	}
+
+	pred := OomkillPredicate{
+		Funcs: predicate.Funcs{
+			UpdateFunc: updateEventFunc,
+		},
+	}
+
+	podSource := source.Kind{Type: &corev1.Pod{}}
+	ctrl.NewControllerManagedBy(mgr).
+		For(&autoscalingv1alpha1.Hvpa{}).
+		Watches(&hvpaSource, &hvpaHandler).
+		Watches(&hpaSource, &hpaHandler).
+		Watches(&vpaSource, &vpaHandler).
+		Watches(&podSource, eventHandler).WithEventFilter(pred).Complete(r)
```

Comment:
Might have to split this into two separate controllers. For example,

```go
ctrl.NewControllerManagedBy(mgr).
		For(&autoscalingv1alpha1.Hvpa{}).
		Watches(&hvpaSource, &hvpaHandler).
		Watches(&hpaSource, &hpaHandler).
		Watches(&vpaSource, &vpaHandler).Complete(r)

podSource := source.Kind{Type: &corev1.Pod{}}
ctrl.NewControllerManagedBy(mgr).
		For(&autoscalingv1alpha1.Hvpa{}).
		Watches(&podSource, eventHandler).WithEventFilter(pred).Complete(r)
```

Otherwise `WithEventFilter(pred)` might apply to all watches.

Thanks @ggaurav10 for pointing this out. 👍

Reply:
In this way the second `.Complete()` would also start second watch on hvpa objects with the predicate `pred`. I am currently looking for a way to use same controller/builder to do this. If there is no way, then we may need to initialise controller separately as is done in v1. WDYT?

Reply:
1. Yes. In the suggestion I made above, there would be two watches on the `HVPA` which is not good.
2. `Owns` then [apparently](https://github.com/gardener/hvpa-controller/blob/de0bceeeab9811b16eb7cec2a03f43c9a3c684f9/vendor/sigs.k8s.io/controller-runtime/pkg/builder/controller.go#L84-L90) `EnqueueRequestForOwner` is automatic. This is best suited for `HPA` and `VPA`.
3. I think it is simpler to handle the other (non-pod) resources gracefully in the `pred`.

```go
podSource := source.Kind{Type: &corev1.Pod{}}
ctrl.NewControllerManagedBy(mgr).
		For(&autoscalingv1alpha1.Hvpa{}).
		Watches(&hvpaSource, &hvpaHandler).
		Owns(&hpaSource, &hpaHandler).
		Owns(&vpaSource, &vpaHandler).
		Watches(&podSource, eventHandler).WithEventFilter(pred).Complete(r)
```

Reply:
So going ahead with 
```
	ctrl.NewControllerManagedBy(mgr).
		For(&autoscalingv1alpha1.Hvpa{}).
		Owns(&autoscaling.HorizontalPodAutoscaler{}).
		Owns(&vpa_api.VerticalPodAutoscaler{}).
		Watches(&podSource, eventHandler).WithEventFilter(pred).Complete(r)
```
The predicate will try to cast it to type `Pod`, and return `false` if it fails.

Reply:
> The predicate will try to cast it to type Pod, and return false if it fails.

 No. That won't work. It should check if the type is `HVPA`, `HPA` or `VPA` and if so, return `true`. If the type id `Pod` then it can go ahead with existing logic.

You might also have to introduce dummy funcs for `Create`, `Delete` etc. Otherwise those events for `HVPA` etc. also will be filtered out.

Reply:
yes. was just about to update the thread with this.
Thanks for capturing. :)

Reply:
Wait a min, the dummy funcs part is already handled in [`Funcs`](https://github.com/gardener/hvpa-controller/blob/de0bceeeab9811b16eb7cec2a03f43c9a3c684f9//vendor/sigs.k8s.io/controller-runtime/pkg/predicate/predicate.go#L64).

Reply:
I think only change we would need is in [updateEventFunc](https://github.com/ashwani2k/hvpa-controller/blob/migrate-to-latest-kubebuilder/controllers/hvpa_controller.go#L71). If the casting to `Pod` fails, then return `true` (to reconcile). If it is not an update of Pod, then it is and update event of HPA/VPA/HVPA. Rest of the handler code should take care of rest. wdyt?

Reply:
:-1: Sounds good.